src/Microsoft.Diagnostics.Monitoring.WebApi/Metrics/MetricsService.cs

```diff
@@ -41,37 +46,104 @@ protected override async Task ExecuteAsync(CancellationToken stoppingToken)
+            counterSettings.UseSharedSession = process.EndpointInfo.RuntimeVersion?.Major >= 8;
 
-                    _counterPipeline = new MetricsPipeline(client, counterSettings, loggers: new[] { new MetricsLogger(_store.MetricsStore) });
+            _counterPipeline = new MetricsPipeline(client, counterSettings, loggers: new[] { new MetricsLogger(metricsStore) });
 
-                    using var linkedTokenSource = CancellationTokenSource.CreateLinkedTokenSource(stoppingToken, optionsTokenSource.Token);
-                    await _counterPipeline.RunAsync(linkedTokenSource.Token);
+            using var linkedTokenSource = CancellationTokenSource.CreateLinkedTokenSource(stoppingToken, optionsTokenSource.Token);
+            await _counterPipeline.RunAsync(linkedTokenSource.Token);
+        }
+
+        private async Task GetMetricsFromMultipleProcessesAsync(CancellationToken stoppingToken)
+        {
+            var trackedProcesses = new Dictionary<int, Task>();
+            while (!stoppingToken.IsCancellationRequested)
+            {
+                stoppingToken.ThrowIfCancellationRequested();
+
+                try
+                {
+                    DiagProcessFilter filter = DiagProcessFilter.FromConfiguration(_processFilterMonitor.CurrentValue);
+                    IEnumerable<IProcessInfo> processes = await _services.GetProcessesAsync(filter, stoppingToken);
+
+                    foreach (var process in processes)
+                    {
+                        int processId = process.EndpointInfo.ProcessId;
+                        if (!trackedProcesses.ContainsKey(processId))
+                        {
+                            MetricsStore metricsStore = _store.GetOrCreateStoreFor(process);
+                            var processMetrics = new SingleProcessMetricsService(_optionsMonitor,
+                                _counterOptions.CurrentValue, process, metricsStore);
+                            trackedProcesses.Add(processId, processMetrics.StartMetricsPipelineForProcessAsync(stoppingToken));
+                        }
+                    }
+
+                    Task delay = Task.Delay(5000, stoppingToken);
+                    Task completedTask = await Task.WhenAny(trackedProcesses.Values.Concat(new [] {delay}));
+                    if (completedTask != delay)
+                    {
+                        SingleProcessMetricsService[] allCompleted = trackedProcesses.Values
+                            .OfType<Task<SingleProcessMetricsService>>()
+                            .Where(task => task.IsCompleted)
+                            .Select(task => task.Result)
+                            .ToArray();
+
+                        foreach (var completed in allCompleted)
+                        {
+                            _store.RemoveMetricsForPid(completed.ProcessId);
```

Comment:
Rather than keeping a list of tasks and checking their results, you could create continuations to StartMetricsPipelineForProcess that remove the pid from a set and remove the metrics from the store.